crypto/src/hash/rpo/mod.rs

```diff
@@ -0,0 +1,925 @@
+use super::{ElementHasher, HashFn};
+use crate::{Felt, FieldElement, StarkField, ONE, ZERO};
+use core::{convert::TryInto, ops::Range};
+
+mod digest;
+pub use digest::RpoDigest256;
+
+#[cfg(test)]
+mod tests;
+
+mod mds_freq;
+use mds_freq::mds_multiply_freq;
+
+// CONSTANTS
+// ================================================================================================
+
+/// Sponge state is set to 12 field elements or 96 bytes; 8 elements are reserved for rate and
+/// the remaining 4 elements are reserved for capacity.
+const STATE_WIDTH: usize = 12;
+
+/// The rate portion of the state is located in elements 4 through 11.
+const RATE_RANGE: Range<usize> = 4..12;
+const RATE_WIDTH: usize = RATE_RANGE.end - RATE_RANGE.start;
+
+const INPUT1_RANGE: Range<usize> = 4..8;
+const INPUT2_RANGE: Range<usize> = 8..12;
+
+/// The capacity portion of the state is located in elements 0, 1, 2, and 3.
+const CAPACITY_RANGE: Range<usize> = 0..4;
+
+/// The output of the hash function is a digest which consists of 4 field elements or 32 bytes.
+///
+/// The digest is returned from state elements 4, 5, 6, and 7 (the first four elements of the
+/// rate portion).
+const DIGEST_RANGE: Range<usize> = 4..8;
+const DIGEST_SIZE: usize = DIGEST_RANGE.end - DIGEST_RANGE.start;
+
+/// The number of rounds is set to 7 to target 128-bit security level
+const NUM_ROUNDS: usize = 7;
```

Comment:
Most likely I'm miscalculating something, but just to double check.

If we perform a Gröbner basis attack, we have that

`d = state width · (alpha - 1) · (rounds - 1) / 2 + 2 = 12 · (7 - 1) · (7 - 1) / 2 + 2 = 218`
`vmin = state width · (rounds - 1) + rate = 12 · (7 - 1) + 4 = 76`
`vmax = state width · (rounds - 1) + rate = 12 · (7 - 1) + 11 = 83` (using exclusive range up to `12`)

Then we have a minimum number of rounds set to `8` to satisfy that security level. Did I miss something?

I built this small sage script to demonstrate the calculation
https://sagecell.sagemath.org/?z=eJxdUVtuwyAQ_LfkO8wnbp2qIFmtIrVHyBESrTFOUTFEBFvt7buYJG2zEmgfM7O7MJgRmpyeHSVziGH2wxkCE1pOn0jb9M3u2eg5sntwZjGOE-ROH4QG27oCW2Qy3pi1ubFKYdDBc8HR1A-EHbYYXQgR4vmpwwNE0dnIhoMpJ3Zr8AjVFIHlP5sxAldM7lpQieLRJIaq_d2spT5ySydhPXP80UBI3kF1t_mz2RG99WGy5PgBFj7MyG3WHS5R-1tosFd4v7b-I5Stj4Y-L29j0hw9tLGsK9e1J_qC6FisCNVVXd3_gZCqxWsLqfh6aX4AmVhu9A==&lang=sage&interacts=eJyLjgUAARUAuQ==

Reply:
Note: its most likely related to the rounding used in the script, but just double checking

Reply:
This implementation is based on variant of Rescue described here: https://github.com/ASDiscreteMathematics/rpo and the motivation for setting number of rounds to 7 is described in section 4.2 [here](https://github.com/ASDiscreteMathematics/rpo/blob/master/report/report.pdf).